mailbox/tool/src/main/java/org/apache/james/mailbox/indexer/ReIndexerImpl.java

```diff
@@ -0,0 +1,138 @@
+    }
+
+
+    public synchronized void reIndex() throws MailboxException {
+        MailboxSession mailboxSession = mailboxManager.createSystemSession("re-indexing", LOGGER);
+        List<MailboxPath> mailboxPaths = mailboxManager.list(mailboxSession);
+        GlobalRegistration globalRegistration = new GlobalRegistration();
+        mailboxManager.addGlobalListener(globalRegistration, mailboxSession);
+        for (MailboxPath mailboxPath : mailboxPaths) {
+            if (globalRegistration.indexThisPath(mailboxPath)) {
+                reIndex(mailboxPath, mailboxSession);
+            }
+        }
+        mailboxManager.removeGlobalListener(globalRegistration, mailboxSession);
+    }
+
+    private void reIndex(MailboxPath path, MailboxSession mailboxSession) throws MailboxException {
+        MailboxRegistration mailboxRegistration = new MailboxRegistration(path);
+        mailboxManager.addListener(path, mailboxRegistration, mailboxSession);
+        LOGGER.info("Intend to reindex " + path);
+        Mailbox<Id> mailbox = mailboxSessionMapperFactory.getMailboxMapper(mailboxSession).findMailboxByPath(path);
+        messageSearchIndex.delete(mailboxSession, mailbox, MessageRange.all());
+        handleIterations(mailboxSession,
+            mailboxRegistration,
+            mailbox,
+            mailboxSessionMapperFactory.getMessageMapper(mailboxSession)
+                .findInMailbox(mailbox,
+                    MessageRange.all(),
+                    MessageMapper.FetchType.Full,
+                    LIMIT));
+        mailboxManager.removeListener(path, mailboxRegistration, mailboxSession);
+        LOGGER.info("Finish to reindex " + path);
+    }
+
+    private void handleIterations(MailboxSession mailboxSession, MailboxRegistration mailboxRegistration, Mailbox<Id> mailbox, Iterator<Message<Id>> iterator) throws MailboxException {
+        while (iterator.hasNext()) {
+            Message<Id> message = iterator.next();
+            ImpactingMessageEvent impactingMessageEvent = findMostRelevant(mailboxRegistration.getImpactingEvents(message.getUid()));
+            if (impactingMessageEvent == null) {
+                messageSearchIndex.add(mailboxSession, mailbox, message);
+            } else if (impactingMessageEvent instanceof FlagsMessageEvent) {
+                message.setFlags(((FlagsMessageEvent) impactingMessageEvent).getFlags());
+                messageSearchIndex.add(mailboxSession, mailbox, message);
+            }
+        }
+    }
+
+    private ImpactingMessageEvent findMostRelevant(Collection<ImpactingMessageEvent> messageEvents) {
+        if (messageEvents == null) {
+            return null;
+        }
+        ImpactingMessageEvent last = null;
+        for (ImpactingMessageEvent impactingMessageEvent : messageEvents) {
+            if (impactingMessageEvent.getType().equals(ImpactingEventType.Deletion)) {
+                return impactingMessageEvent;
+            }
+            last = impactingMessageEvent;
```

Comment:
if you get a list of messageEvents, you can just split your "find deletion" loop from the getLast by using Iterables.getLast()